program/rust/src/validator.rs

```diff
@@ -0,0 +1,141 @@
+            slot,
+            timestamp,
+        )
+    };
+
+    // If the aggregate was successfully updated, calculate the difference and update TWAP.
+    if updated {
+        let agg_diff = (slot as i64) - price_account.prev_slot_ as i64;
+
+        // See comment on unsafe `c_upd_aggregate` call above for details.
+        unsafe {
+            c_upd_twap(price_account as *mut PriceAccount as *mut u8, agg_diff);
+        }
+
+        // We want to send a message every time the aggregate price updates. However, during the migration,
+        // not every publisher will necessarily provide the accumulator accounts. The message_sent_ flag
+        // ensures that after every aggregate update, the next publisher who provides the accumulator accounts
+        // will send the message.
+        price_account.message_sent_ = 0;
+        if price_account.update_price_cumulative().is_err() {
+            // Revert in case of a bad trading status.
+            return false;
+        }
+    }
+
+    true
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum AggregationError {
+    #[error("NotPriceFeedAccount")]
+    NotPriceFeedAccount,
+    #[error("V1AggregationMode")]
+    V1AggregationMode,
+    #[error("AlreadyAggregated")]
+    AlreadyAggregated,
+}
+
+#[derive(Debug)]
+pub struct AggregationOutcome {
+    /// Messages that should be included in the merkle tree
+    /// (a price feed message and a TWAP message).
+    pub messages: [Vec<u8>; 2],
+    /// A flag indicating that the changes to `price_account_data`
+    /// should be saved to the price account.
+    pub commit:   bool,
+}
+
+/// Attempts to read a price account and create a new price aggregate if v2
+/// aggregation is enabled on this price account.
+/// `price_account_data` will be modified in case of successful aggregation.
+/// However, the changes to `price_account_data` should not be stored in the
+/// account if `commit = false` in the returned value or if an error is returned.
+/// Note that the `messages` will be returned whenever possible, even if
+/// aggregation fails for some reason.
+pub fn aggregate_price(
+    slot: u64,
+    timestamp: i64,
+    price_account_pubkey: &Pubkey,
+    price_account_data: &mut [u8],
+) -> Result<AggregationOutcome, AggregationError> {
+    let price_account = validate_price_account(price_account_data)?;
+    if price_account.agg_.pub_slot_ == slot {
+        // Avoid v2 aggregation if v1 aggregation has happened in the same slot
+        // (this should normally happen only in the slot that contains the v1->v2 transition).
+        return Err(AggregationError::AlreadyAggregated);
+    }
+    let commit = update_aggregate(slot, timestamp, price_account);
+    let messages = [
+        price_account
+            .as_price_feed_message(price_account_pubkey)
+            .to_bytes(),
+        price_account
+            .as_twap_message(price_account_pubkey)
+            .to_bytes(),
+    ];
```

Comment:
I think we need to add a check to not add it if the message buffer is not cleared. It can result it not sending the same account state temporarily until a new tx to update price comes but i think it's better than having duplicate messages.